pkg/reconciler/pipelinerun/pipelinerun.go

```diff
@@ -597,16 +622,49 @@ func (c *Reconciler) reconcile(ctx context.Context, pr *v1alpha1.PipelineRun) er
 			}
 		}
 	}
-	before := pr.Status.GetCondition(apis.ConditionSucceeded)
-	after := resources.GetPipelineConditionStatus(pr, pipelineState, c.Logger, d)
-	pr.Status.SetCondition(after)
-	reconciler.EmitEvent(c.Recorder, before, after, pr)
-
+	pipelineTasksStatusBefore := pr.Status.GetCondition(apis.ConditionSucceeded)
+	pipelineTasksStatusAfter := resources.GetPipelineConditionStatus(pr, pipelineTasks, c.Logger, d)
+	// begin executing final tasks when all non-final tasks are completed successfully or any one of them have failed
+	if pipelineTasksStatusAfter.IsTrue() || (pipelineTasksStatusAfter.IsFalse() && pipelineTasksStatusAfter.Reason == resources.ReasonFailed) {
+		if err := c.executeFinalTasks(pipelineState, finalTasks, pr, pipelineSpec, pipelineTasksStatusAfter, as.StorageBasePath(pr)); err != nil {
+			return err
+		}
+	} else {
+		// set the PipelineRun Status if still executing non-final tasks
+		pr.Status.SetCondition(pipelineTasksStatusAfter)
+		reconciler.EmitEvent(c.Recorder, pipelineTasksStatusBefore, pipelineTasksStatusAfter, pr)
+	}
 	pr.Status.TaskRuns = getTaskRunsStatus(pr, pipelineState)
 	c.Logger.Infof("PipelineRun %s status is being set to %s", pr.Name, pr.Status.GetCondition(apis.ConditionSucceeded))
 	return nil
 }
 
+func (c *Reconciler) executeFinalTasks(pipelineState resources.PipelineRunState, finalTasks resources.PipelineRunState, pr *v1alpha1.PipelineRun, pipelineSpec *v1alpha1.PipelineSpec, pipelineTasksStatusAfter *apis.Condition, storageBasePath string) error {
```

Comment:
when possible we should try to avoid putting functions on the reconciler - its easier to test and maintain over time if we put them in another package and pass in any attributes of the reconciler they need